### ISSUE TITLE
TCGA Workflow: Analyze cancer genomics and epigenomics data using Bioconductor packages (workflow package)

### ISSUE THREAD

**Issue by tiagochst** [Bioconductor/Contributions#229]
Update the following URL to point to the GitHub repository of
the package you wish to submit to _Bioconductor_

- Repository: https://github.com/BioinformaticsFMRP/TCGAWorkflow

Confirm the following by editing each check box to '[x]'

- [X] I understand that by submitting my package to _Bioconductor_,
  the package source and all review commentary are visible to the
  general public.

- [X] I have read the _Bioconductor_ [Package Submission][2]
  instructions. My package is consistent with the _Bioconductor_
  [Package Guidelines][1].

- [X] I understand that a minimum requirement for package acceptance 
  is to pass R CMD check and R CMD BiocCheck with no ERROR or WARNINGS. 
  Passing these checks does not result in automatic acceptance. The 
  package will then undergo a formal review and recommendations for 
  acceptance regarding other Bioconductor standards will be addressed.

- [X] My package addresses statistical or bioinformatic issues related
  to the analysis and comprehension of high throughput genomic data.

- [X] I am committed to the long-term maintenance of my package. This
  includes monitoring the [support site][3] for issues that users may
  have, subscribing to the [bioc-devel][4] mailing list to stay aware
  of developments in the _Bioconductor_ community, responding promptly
  to requests for updates from the Core team in response to changes in
  _R_ or underlying software.

I am familiar with the essential aspects of _Bioconductor_ software
management, including:

- [X] The 'devel' branch for new packages and features.
- [X] The stable 'release' branch, made available every six
      months, for bug fixes.
- [X] _Bioconductor_ version control using [Subversion][5]
  (optionally [via GitHub][6]).

For help with submitting your package, please subscribe and post questions
to the [bioc-devel][4] mailing list.

[1]: https://bioconductor.org/developers/package-guidelines/
[2]: https://bioconductor.org/developers/package-submission/
[3]: https://support.bioconductor.org
[4]: https://stat.ethz.ch/mailman/listinfo/bioc-devel
[5]: http://bioconductor.org/developers/how-to/source-control/
[6]: http://bioconductor.org/developers/how-to/git-mirrors/

**Comment by vobencha** [comment]
Hi,

The Depends field in DESCRIPTION should list the packages from which you call functions or create classes. Looks like it might boil down to just these 
library(TCGAbiolinks)
library(stringr)
library(ComplexHeatmap)
library(SummarizedExperiment

The rest can go in Suggests. Please add a NAMESPACE to import the packages in Depends.

Valerie

**Comment by tiagochst** [comment]
@vobencha Hi, thank you for the review, I added the namespace. I'm sorry if I understood it wrong, but all the Depends have at least one function being used. I added those functions to the package [NAMESPACE](https://github.com/BioinformaticsFMRP/f1000_TCGA_Workflow/blob/master/NAMESPACE).

**Comment by vobencha** [comment]
OK, slight change of plans. These 'workflows as packages' are still somewhat new so we are figuring the process out. Since you don't have any internal R code the packages are only needed for the vignette. Please move the list of packages to 'Suggests' and don't import them in the NAMESPACE. You should still have a NAMESPACE but for now it will be empty.

Thanks.
Valerie

**Comment by tiagochst** [comment]
Hi Valerie,

It is done.

Best regards,
Tiago

**Comment by vobencha** [comment]
Please see comments and suggestions below.

1) R/ vs vignette/ directory

A workflow package usually doesn't define new R functions but reuses from existing packages. Defining little functions used in *apply() loops are fine but the vignette is not the place to define substantial helpers. If you need substantial helper functions they should be in R/, not the vignette. Please move the functions from the "internal functions" section in the vignette to R/utilities.R. Also add a man page, man/utilities.Rd, that briefly documents them and export them in the NAMESPACE. The functions they depend on will need to be moved from Suggests to Depends or Imports and import() or importFrom() in the NAMESPACE.

I think all of these should be moved to utilities.R:

~/proj/git/software/f1000_TCGA_Workflow/vignettes >grep "<- function" *
f1000.Rmd:matched_met_exp <- function(project, n = NULL){
f1000.Rmd:gaiaCNVplot <- function (calls, cancer=NULL, threshold=0.01)
f1000.Rmd:gaiaCNVplot <- function (calls, cancer=NULL, threshold=0.01)
f1000.Rmd:get.adjacency.biogrid <- function(tmp.biogrid, names.genes = NULL){
f1000.Rmd:matched_met_exp <- function(project, n = NULL){
f1000.Rmd:matched_met_exp <- function(project, n = NULL){
f1000.Rmd:TCGAsurvival_TFplot <- function(TF,mee, clinical, percentage = 0.3){
f1000.Rmd:TCGAsurvival_TFplot <- function(TF,mee, clinical, percentage = 0.3){

Sorry for missing this on the first review but the extra code really clutters the vignette and should be split out.

2) R CMD build

I cannot build the package:

```
~/proj/git/software >R-dev CMD build f1000_TCGA_Workflow/
* checking for file ‘f1000_TCGA_Workflow/DESCRIPTION’ ... OK
* preparing ‘TCGAWorkflow’:
* checking DESCRIPTION meta-information ... OK
* installing the package to build vignettes
* creating vignettes ... ERROR
See system.file("LICENSE", package="MotifDb") for use restrictions.
Download DNA methylation information
Accessing GDC. This might take a while...
Download gene expression information
Accessing GDC. This might take a while...
Download DNA methylation information
Accessing GDC. This might take a while...
Download gene expression information
Accessing GDC. This might take a while...
Accessing GDC. This might take a while...
GDCdownload will download 10 files. A total of 212.858413 MB
Downloading as: Thu_Jan_12_06_28_07_2017.tar.gz
Downloading genome information. Using: Homo sapiens genes (GRCh37.p13)
Quitting from lines 54-235 (f1000.Rmd)
Error: processing vignette 'f1000.Rmd' failed with diagnostics:
Request to BioMart web service failed. Verify if you are still connected to the internet.  Alternatively the BioMart web service is temporarily down.
Execution halted
```

The calls to BioMart (or any web service) must be wrapped in tryCatch() and given a limited number of tries. Please look at the GEOsearch/R/utilities.R file. Martin recently helped that author design functions that read / access urls with a timeout. Such precautions can prevent the package from putting unnecessary load on the build machines. These also flag web sites that are unreliable. Please wrap your download calls in similar helper functions. If after 3 tries you can't get the file, either error out or move forward in the workflow without that data.

3) loading packages

Load each package once. If they are loaded at the top of the vignette they do not need to be loaded again in separate sections. For example, this is the number of times the TCGAbiolinks package is loaded:

~/f1000_TCGA_Workflow/vignettes >grep "library(TCGAbiolinks)" *
f1000.Rmd:    library(TCGAbiolinks)
f1000.Rmd:library(TCGAbiolinks)
f1000.Rmd:library(TCGAbiolinks)
f1000.Rmd:library(TCGAbiolinks)
f1000.Rmd:library(TCGAbiolinks)
f1000.Rmd:library(TCGAbiolinks)
f1000.Rmd:library(TCGAbiolinks)
f1000.Rmd:library(TCGAbiolinks)
f1000.Rmd:library(TCGAbiolinks)
f1000.Rmd:library(TCGAbiolinks)
f1000.Rmd:library(TCGAbiolinks)
f1000.Rmd:library(TCGAbiolinks)
f1000.Rmd:library(TCGAbiolinks)
f1000.Rmd:library(TCGAbiolinks)

4) Direct slot access
Accessing class slots with '@' goes against best practices. There are many reasons for this - slot names can change, the class can be invalidated if the wrong data is put in the slot etc. Instead you should use getter and setters to access and change the data. Do these classes not have accessor functions? Please explain why you are using '@' throughout the code.

Please do NOT encourage the user to use '@':
```
# To access the data you should use the getData function
# or simply access with @ (for example gbm.data@Clinical)
gbm.mut <- getData(gbm.data,"Mutations")
gbm.clin <- getData(gbm.data,"Clinical")
gbm.gistic <- getData(gbm.data,"GISTIC")
```
Does this class not have accessors?
```
# get GISTIC results
gistic.allbygene <- gistic@GISTIC@AllByGene
gistic.thresholedbygene <- gistic@GISTIC@ThresholedByGene
```

5) I would reword this to say the workflow 'makes use' of these packages, not 'introduces' them.
```
This workflow introduces the following Bioconductor packages:

* [AnnotationHub](http://www.bioconductor.org/packages/AnnotationHub/)
* [ChIPSeeker](http://www.bioconductor.org/packages/ChIPseeker/)
* [ComplexHeatmap](http://www.bioconductor.org/packages/ComplexHeatmap/)
* [pathview](http://www.bioconductor.org/packages/pathview/)
* [ELMER](http://www.bioconductor.org/packages/ELMER/)
* [GAIA](http://www.bioconductor.org/packages/gaia/)
* [MINET](http://www.bioconductor.org/packages/minet/)
* [RTCGAtoolbox](http://www.bioconductor.org/packages/RTCGAtoolbox/)
* [TCGAbiolinks](http://www.bioconductor.org/packages/TCGAbiolinks/)
* [AnnotationHub](http://www.bioconductor.org/packages/GAIA/)
```
Also, this is not a complete list of packages used in the vignette. It will be difficult to keep this up to date. I would suggest removing this section and just cite or talk about the packages as you go along in the vignette. Alternatively, you could mention once, at the top of the vignette that package landing pages can be found through the biocViews interface here:

http://www.bioconductor.org/packages/release/BiocViews.html#___Software

Valerie

**Comment by tiagochst** [comment]
Thank you for the review. I'll work on those points next week. I'll let you
know when I fix them.


Tiago Chedraoui Silva

**Comment by tiagochst** [comment]
@vobencha I updated the workflow

- [X] 1)  Functions were moved to R/utilities.R and their dependencies were moved to the import field in the DESCRIPTION file
- [X] 2)  The function that will be executed to compile the vignette has the 3 tries. If there is more than 3 tries there will be an error. I followed  GEOsearch/R/utilities.R as suggested. This is the [https://github.com/BioinformaticsFMRP/TCGAbiolinks/commit/985ef8dfaa39bc8f701c2576ef84f5c3a6a4172f](commit)
- [X] 3) I removed the excessive number of libraries. Now it should have one of each only (at least the ones that will be executed in the code)
- [X] 4) I removed the @, now I'm using the accessor. For the mee@sample$TN, there is no setter. I'm rewriting ELMER package. So this will be removed latter from this workflow when I update ELMER to the new version.
- [X] 5) I removed the package list and added a link to biocViews interface

**Comment by vobencha** [comment]
Hi,
Thanks for making the changes. A few more comments below. The most important is the amount of data being downloaded and processed (#9).

6. Inconsistent names
The github repo is f1000_TCGA_Workflow, the title in DESCRIPTION is TCGAWorkflow and the title of the vignette is f1000.Rnw. I'd suggest making a more descriptive name for the vigentte, if nothing else have it match the title of the package, e.g., TCGAWorkflow.Rnw.

7. export functions used in the vignette
Functions defined in the R/ directory will need to be exported for use in the vignette. Please add the exported functions to the NAMESPACE and load the package at the top of the vignette, e.g., library(TCGAWorkflow).

8. no output from the single package builder
You may have noticed this ... ? I think the package is not completing R CMD build. Have you run R CMD build and R CMD check? I tried running R CMD build and it was still running after 50 minutes.

9. volume of downloads
The show stopper is the amount of data you are trying to download a process. The vignette tries to download > 600 files totaling > 2000 MB. Here is just part of the output from running the vignette before I killed it:
```
GDCdownload will download 10 files. A total of 212.965792 MB
Downloading as: Thu_Jan_19_10_10_55_2017.tar.gz
...
Of the 156 files for download 10 already exist.
We will download only those that are missing ones.
GDCdownload will download 146 files. A total of 221.007115 MB
...
Of the 516 files for download 10 already exist.
We will download only those that are missing ones.
GDCdownload will download 506 files. A total of 768.699866 MB
```
and the list goes on.

Unfortunately this is not reasonable (in time or space) to run on our build machines and I can't imagine it would be easy for a user to reproduce. Can you find similar data in one of the Bioconductor repos? There are quite a few TCGA data sets in the ExperimentalData repo (type 'TCGA' in search box).

http://www.bioconductor.org/packages/release/BiocViews.html#___AnnotationData
http://www.bioconductor.org/packages/release/BiocViews.html#___ExperimentData

If not, you could create your own experimental data package that can be reused in the workflow package.

Valerie

**Comment by tiagochst** [comment]
@vobencha  Hi,

- [x] 6. I changed all the names to TCGAWorkflow
- [x] 7. The functions are now exported in the NAMESPACE
- [x] 8. The build was taking one hour, I reduced the data downloaded and R CMD build and R CMD check are Ok, I still receive some NOTES. It is running on my machine all the steps in 30 min.
- [x] 9. I agree with you. I decided to add to the package a subset of data necessary to execute the steps. Unfortunately, the data available in Bioconductor would still give the same problem of space requirement, the user would need to download more than 500MB. I also reduced the number data to be downloaded (in case the user wants to download). 

What do you think about this approach, do you believe this would help the workflow to be easier ? Unfortunately, we would need to have more samples to have significant results. But I agree it should be faster.

**Comment by tiagochst** [comment]
By the way, when I changed the name of the repository the commit are not linked to this issue.
How should I proceed to fix it?
The last version is in the [github repository](https://github.com/BioinformaticsFMRP/TCGAWorkflow)

**Comment by vobencha** [comment]
I'm cc'ing @mtmorgan to help with this. Martin, how should we proceed when the repo has changed names? Initially I thought we needed to close this issue and open another but I think that may not be the case.
Val

**Comment by mtmorgan** [comment]
I think the link in the original comment that opened the issue can be edited to point to the new repo.

**Comment by tiagochst** [comment]
I changed the link, but it didn't work. Should I close and open another issue?

**Comment by mtmorgan** [comment]
Does the new repository have the [web hook][] defined? Can you confirm that the web hook fired on version bump -- under 'Settings' on top right of repository page, 'Web hooks', the link should appear, click on 'Edit' and scroll down to top of the hash tags; look for the one fired at the time of the version bump and 'Response' status.

[web hook]: https://github.com/Bioconductor/Contributions#adding-a-web-hook

**Comment by tiagochst** [comment]
<img width="740" alt="screen shot 2017-01-25 at 11 05 42 am" src="https://cloud.githubusercontent.com/assets/145529/22304888/46c0a9f4-e2ee-11e6-9fd2-454e077c9e4a.png">

**Comment by mtmorgan** [comment]
Thanks, this is the first time we've dealt with this; can you press the 'Redeliver' button on the Recent Delivery you show above?

**Comment by tiagochst** [comment]
@mtmorgan  @vobencha Thanks! It worked

**Comment by vobencha** [comment]
Hi,
Responding to comments in https://github.com/Bioconductor/Contributions/issues/229#issuecomment-274941697. Thanks for addressing many of the issues.

#8. 30 minutes is a reasonable amount of time for the workflow to run.
#9. I understand results are more meaningful when more data are downloaded, however, 500 MB is too much. It would be ok to show the full download in the vignette but not evaluate it and say 'instead we will download a subset ... etc.'. Clearly state the full download is >= 500 MB so users without a sufficient internet connection or space are forewarned.

    The data should not be included in the workflow package but should be a separate experimental data package.

Valerie

**Comment by tiagochst** [comment]
OK! I'll create the experimental data package. Just to know, I'll need to create another issue? How do I link the issue with this one?

**Comment by tiagochst** [comment]
I just [found](https://github.com/Bioconductor/Contributions/blob/master/CONTRIBUTING.md#submitting-related-packages) . Sorry

**Comment by tiagochst** [comment]
AdditionalPackage: https://github.com/BioinformaticsFMRP/TCGAWorkflowExampleData

**Comment by vobencha** [comment]
Review of data package:
- Change name to TCGAWorkflowData
- In the Description field of the DESCRIPTION file please say what kind of data these are, e.g., 9 data sets of expression data for cancer XYZ for 100 samples from https://download/site etc.
- Remove 'Date' field from DESCRIPTION; this will be auto generated when the package is built with R CMD build
- The man pages can be separate or all on one page. A single page would probably be more efficient here because all man pages must have
i) more detail about how the data were processed, how the subsets were made etc.
ii) an \example section that shows how to load the data (if only one man page you only need to show how to load a couple of datasets)
iii) define abbreviations such as GBM (what does this mean?)

Here is an example of a complete man page for a data set:
```
> library(pasillaBamSubset)
> ?pasillaBamSubset
```
Note how the name of the data package brings you to one page that lists all data sets - very easy to find.

Valerie

**Comment by vobencha** [comment]
You've done a great job with the vignette in the data package. You could reference users to that on the man page(s) with browseVignettes("TCGAWorkflowData"). The man pages still need a mention of what I outlined above but the bulk of the detail re: creating the datasets can be in the vignette.

**Comment by tiagochst** [comment]
AdditionalPackage: https://github.com/BioinformaticsFMRP/TCGAWorkflowData

**Comment by tiagochst** [comment]
Hi @vobencha, just one question. The workflows accepts any type of Rmd?
I was thinking to update it to use DT package to shows some tables.
Like here: [http://bioinformaticsfmrp.github.io/TCGAbiolinks/query.html](http://bioinformaticsfmrp.github.io/TCGAbiolinks/query.html)

**Comment by vobencha** [comment]
Not sure where we are on this. When you're ready for me to take another look mention me @vobencha.
Valerie

**Comment by tiagochst** [comment]
Hi, sure. I'm working on other projects right now. I should be back on this next week. Once I fix the probles I'll mention you. I have just one question. One of the problems is timeout (
TIMEOUT: R CMD build exceeded 10mins). This limit is new for workflows?

**Comment by tiagochst** [comment]
@vobencha @mtmorgan  The last build report gave timeout for linux and mac (which I tested yesterday and was working), but it was not supposed to give timeout. How do I know if it was build as a workflow?

Thanks!

"Workflow vignettes may deal with larger data sets and/or be more computationally intensive than typical Bioconductor package vignettes. For this reason, the automated builder that produces these vignettes does not have a time limit (in contrast to the Bioconductor package building system which will time out if package building takes too long)."

**Comment by vobencha** [comment]
I'll take another look at this tomorrow.
Valerie

**Comment by vobencha** [comment]
Taigo,

I am not feeling good about this workflow. It's true that workflows are expected to use larger data and run longer than software or experimental data packages this one is really pushing the limits.

## General comments:
I tried to build TCGAWorkflow on my linux laptop with 8 cores and 16 GB RAM and it almost ground to a halt - I had to kill it. This indicates either inefficient code, attempting to read / manipulate too much data or both. I will not approve this package until I can build it on my local system.  I realize you're trying to make code and results available to those who want to follow what was done in the publication. I don't, however, think it's reasonable to expect users to wait >30 minutes for a workflow to complete or to have access to a large computational server.

The point of creating a dedicated data package for the workflow was to (i) avoid downloading large amounts of data and (ii) avoid lengthy processing steps. You have created the TCGAWorkflowData package but are still downloading an large number of files and have long-running analyses. Is it possible to save sub-sample of the downloads in the data package? Is it possible to not evaluate all of the analysis steps? You need to find a middle ground where you can convey you message also let the user move through the workflow in a reasonable time.

## Software package:
1) inefficiencies
- I've noticed these functions take a long time - maybe you can reduce the data or save a subset as I suggest above.
```
getFirehoseData, GDCquery(), GDCquery_maf()
```
- This loop is slow. See Martin's suggestions here for vectorizing, pre-allocating etc.:

http://www.bioconductor.org/developers/how-to/efficient-code/
```
> dim(mat)
[1]  91 447

for (i in 1:nrow(mat)) {
    curGene <- rownames(mat)[i]
    setTxtProgressBar(pb, i)
    for (j in 1:ncol(mat)) {
        curSample <- colnames(mat)[j]

        if (length(intersect(mut$Tumor_Sample_Barcode, curSample))==1){
            mat1 <- mut[mut$Tumor_Sample_Barcode == curSample,]
            if (length(intersect(mat1$Hugo_Symbol, curGene))==1){
                mat3 <- mat1[mat1$Hugo_Symbol == curGene,]
                mat[curGene,curSample]<- as.character(mat3$Variant_Type)[1]
            }
        }
    }
}
```
2) warnings()

When I ran the methods.Rnw vignette interactively there were a number of warnings. Most of these indicate problems with the code and should not be ignored. Because the vignette takes so long to run (and actually does not run on my local machine) we can't see these warnings in the usual output of R CMD build and R CMD check.

Here are a few of the warnings I saw. For the package to be approved these warning must be addressed or explained.

- warning 1
```
> TCGAvisualize_EAbarplot(tf = rownames(ansEA$ResBP),
+                         filename = NULL,
+                         GOBPTab = ansEA$ResBP,
+                         GOCCTab = ansEA$ResCC,
+                         GOMFTab = ansEA$ResMF,
+                         PathTab = ansEA$ResPat,
+                         nRGTab = rownames(dataDEGs),
+                         nBar = 20)
Warning messages:
1: In if (!is.null(GOBPTab) & !is.na(GOBPTab)) { :
  the condition has length > 1 and only the first element will be used
2: In if (!is.null(GOCCTab) & !is.na(GOCCTab)) { :
...
```
- warning 2
```
Writing Output File 'GAIA_GBM_flt.txt' Containing the Significant Regions
File 'GAIA_GBM_flt.txt' Saved
Warning messages:
1: In if (tmp_list[[1]] == -1) { :
  the condition has length > 1 and only the first element will be used
2: In if (tmp_list[[1]] == -1) { :
  the condition has length > 1 and only the first element will be used
```
- warning 3
```
> dataNorm <- TCGAanalyze_Normalization(tabDF = cbind(dataPrep_LGG, dataPrep_GBM),
+                                       geneInfo = geneInfo,
+                                       method = "gcContent") #18323   672
I Need about  11 seconds for this Complete Normalization Upper Quantile  [Processing 80k elements /s] 
Step 1 of 4: newSeqExpressionSet ...
Step 2 of 4: withinLaneNormalization ...
Step 3 of 4: betweenLaneNormalization ...
Step 4 of 4: .quantileNormalization ...
Warning message:
In geneNames[, 1] == names(tmp[which(tmp > 1)]) :
  longer object length is not a multiple of shorter object length
```
3) What are the files in vignette/libs/ used for?
4) What is the purpose of index.Rmd?

There is no need for the loading of 'deps' in that file. Is the concern making sure all dependencies are installed when the package is loaded? If yes, you can move everything that is called with library() in the vignette to Depends in DESCRIPTION instead of Suggests. We have discussed this before - standard software packages list resources used in the vignette and man pages in Suggests. In this case, the vignette is the main attraction so it would be fine to have all in Depends.
These packages would move to Depends:
```
~/TCGAWorkflow/vignettes >grep library methods.Rmd
    library(TCGAWorkflow)
library(TCGAWorkflowData)
library(DT)
library(TCGAbiolinks)
library(TCGAbiolinks)
library(SummarizedExperiment)
library(RTCGAToolbox)
library(TCGAbiolinks)
library(TCGAbiolinks)
library(gaia)
library(GenomicRanges)
library(ComplexHeatmap)
library(circlize)
library(SummarizedExperiment)
...
```
5) Remove the Date fields from DESCRIPTION.
6) Remove all instances of this or explain why you think you need it:
```
R.utils::gcDLLs()
```
Trying to manually manage the loading and unloading of libraries is a tricky business and often better left to R.

7) Remove all of these detach statements:
```
methods.R:detach("package:RTCGAToolbox", unload=TRUE)
methods.R:detach("package:gaia", unload=TRUE)
methods.R:detach("package:minet", unload=TRUE)
methods.R:detach("package:c3net", unload=TRUE)
methods.R:detach("package:pathview", unload=TRUE)
methods.R:detach("package:clusterProfiler", unload=TRUE)
methods.R:detach("package:motifStack", unload=TRUE)
methods.R:detach("package:MotIV", unload=TRUE)
methods.R:detach("package:rGADEM", unload=TRUE)
methods.Rmd:detach("package:RTCGAToolbox", unload=TRUE)
methods.Rmd:detach("package:gaia", unload=TRUE)
methods.Rmd:detach("package:minet", unload=TRUE)
methods.Rmd:detach("package:c3net", unload=TRUE)
methods.Rmd:detach("package:pathview", unload=TRUE)
methods.Rmd:detach("package:clusterProfiler", unload=TRUE)
methods.Rmd:detach("package:motifStack", unload=TRUE)
methods.Rmd:detach("package:MotIV", unload=TRUE)
methods.Rmd:detach("package:rGADEM", unload=TRUE)
```
The warnings() indicate they serve no purpose because most can't be unloaded:
```
warnings()
...
14: ‘minet’ namespace cannot be unloaded:
  namespace ‘minet’ is imported by ‘TCGAbiolinks’ so cannot be unloaded
15: ‘c3net’ namespace cannot be unloaded:
  namespace ‘c3net’ is imported by ‘TCGAbiolinks’ so cannot be unloaded
16: ‘pathview’ namespace cannot be unloaded:
  namespace ‘pathview’ is imported by ‘TCGAbiolinks’ so cannot be unloaded
17: ‘clusterProfiler’ namespace cannot be unloaded:
  namespace ‘clusterProfiler’ is imported by ‘TCGAbiolinks’ so cannot be unloaded
...
```
8) Name the exported functions in a consistent manner. Use either camelCase, dots '.' or underscores, not a mixture.
```
export(gaiaCNVplot)
export(get.adjacency.biogrid)
export(matched_met_exp)
```
9) What is the point of R/TCGAWorkflow.R? It documents TCGAbiolinks, not the workflow.
10) This warning states the data are too big for client-side DataTables and must be fixed:
```
> ## ----echo = TRUE, message = FALSE, warning = FALSE-----------------------
> datatable(clinical, options = list(scrollX = TRUE, keys = TRUE), rownames = FALSE)
Warning message:
In instance$preRenderHook(instance) :
  It seems your data is too big for client-side DataTables. You may consider server-side processing: http://rstudio.github.io/DT/server.html
```

## TCGAWorkflowData package:

You don't need to list all datasets in the Description field of the DESCRIPTION file. It would be fine to say the package contains 7 data sets necessary to follow the TCGAWorkflow ... You have a good man page the user can find easily with ?TCGAWorkflowData that sufficiently describes all the datasets.

Valerie

**Comment by tiagochst** [comment]
Hi Valerie,

I agree with you, I'll keep a complete version in github and I'll create a small/faster version for Biocondutor.

1) I will remove: getFirehoseData, GDCquery(), GDCquery_maf()
2) I'll improve the loop

3) and 4.

I was adding a floating toc: http://bioinformaticsfmrp.github.io/TCGAWorkflow/methods.html
But I can remove it, if it is not accepted.  I was just trying to make it more organized.

4) and  6. and 7.

I added R.utils::gcDLLs() and moved the packages to  Depends in DESCRIPTION, because I reached R limit and I thought the time limit was related to it. This is the problem: http://r.789695.n4.nabble.com/Maximum-of-DLLs-reached-or-how-to-clean-up-after-yourself-td4724655.html

I'll fix it 5, 8,9 and 10. I'll check which warnings I can remove from 3.

I'll come back to you once I fixed it.

Best regards,
Tiago

**Comment by vobencha** [comment]
Comments in https://github.com/Bioconductor/Contributions/issues/229#issuecomment-284028696 sound good, thank you.

Valerie

**Comment by tiagochst** [comment]
@vobencha I fixed all problems except the warnings (I believe 1 and 3 are fixed with the new TCGAbiolinks version) but 2 is from gaia and I'm not the maintainer. 

It is running quite fast now, but I still have timeouts.
<img width="326" alt="screen shot 2017-03-10 at 11 13 16 am" src="https://cloud.githubusercontent.com/assets/145529/23809451/9c8fde5c-0582-11e7-8d28-fe9cd4bcd0aa.png">

**Comment by vobencha** [comment]
Hi,
Thanks for making the changes. As for the timeouts, we've decided to give workflow packages a 30 minute limit on the SPB. To enable this please add this line to DESCRIPTION:
```
Workflow: True
```
Once that's in DESCRIPTION and you have the time under 30 minutes we should see a build.
Valerie

**Comment by vobencha** [comment]
Hi,

- workflow functions
Are you sure the functions you define in the workflow package would not be more appropriate to include in a software package? I ask because workflows are usually light wrapper around multiple software and data packages. In general they don't define their own functions or supply their own data. I asked you to create these functions because the originally the vignette contained too much code. Now that things are cleaned up you might consider adding some (all?) of these to TCGAbiolinks. It seems as though they would be useful to other people ... ?

- docs/ folder
I'm not sure what you want to accomplish with the docs/ folder. Do you mean inst/doc ... where you can put index.html? If yes, you need to add the inst/ folder and a doc/ folder under it.

  inst/doc/ generally has index.html but not much else. Why the .R and .Rmd files? Is the vignette in vignettes/ not the complete document?
```
~/proj/git/software/TCGAWorkflow >ls docs
conclusion.html  conclusion.Rmd  index.R    methods.html  methods.Rmd
conclusion.R     index.html      index.Rmd  methods.R
```
Also, the dates in the html files are old. It is not a good idea to have static dates there - dates should be generated on-the-fly in the vignette code.

Valerie

**Comment by tiagochst** [comment]
Hi,

* docs/ folder
   I used /docs for github, I had a copy in a another format. But as it was old I removed it

* workflow functions
  I moved the functions to TCGAbiolinks, but the CNV plot would be better if added to the gaia package.  Unfortunately, I'm not the maintainer. So, I'm adding to TCGAbiolinks.

Best regards,
Tiago

**Comment by vobencha** [comment]
The Biobase error on windows is our problem.

When I run R CMD build on the workflow I get this error - which is an undeclared dependency since it's not in DESCRIPTION:
```
Loading required package: seqLogo
Quitting from lines 1621-1651 (TCGAWorkflow.Rmd) 
Error: processing vignette 'TCGAWorkflow.Rmd' failed with diagnostics:
there is no package called 'BSgenome.Hsapiens.UCSC.hg19'
```
Please make sure all of these are mentioned in some field in DESCRIPTION:
```
~/proj/git/software/TCGAWorkflow/vignettes >grep library *
TCGAWorkflow.Rmd:    library(TCGAWorkflow)
TCGAWorkflow.Rmd:library(TCGAWorkflowData)
TCGAWorkflow.Rmd:library(DT)
TCGAWorkflow.Rmd:library(TCGAbiolinks)
TCGAWorkflow.Rmd:library(TCGAbiolinks)
TCGAWorkflow.Rmd:library(SummarizedExperiment)
TCGAWorkflow.Rmd:library(RTCGAToolbox)
TCGAWorkflow.Rmd:library(TCGAbiolinks)
TCGAWorkflow.Rmd:library(gaia)
TCGAWorkflow.Rmd:library(GenomicRanges)
TCGAWorkflow.Rmd:library(GenomicRanges)
TCGAWorkflow.Rmd:library(ComplexHeatmap)
TCGAWorkflow.Rmd:library(circlize)
TCGAWorkflow.Rmd:library(SummarizedExperiment)
TCGAWorkflow.Rmd:library(clusterProfiler)
TCGAWorkflow.Rmd:library(pathview)
TCGAWorkflow.Rmd:library(minet)
TCGAWorkflow.Rmd:library(c3net)
TCGAWorkflow.Rmd:library(ComplexHeatmap)
TCGAWorkflow.Rmd:library(rGADEM)
TCGAWorkflow.Rmd:library(BSgenome.Hsapiens.UCSC.hg19)
TCGAWorkflow.Rmd:library(motifStack)
TCGAWorkflow.Rmd:library(MotIV)
TCGAWorkflow.Rmd:library(ChIPseeker)
TCGAWorkflow.Rmd:library(pbapply)
TCGAWorkflow.Rmd:library(ggplot2)
TCGAWorkflow.Rmd:library(AnnotationHub)
TCGAWorkflow.Rmd:library(ggthemes)
TCGAWorkflow.Rmd:library(TCGAbiolinks)
TCGAWorkflow.Rmd:library(ELMER)
```
Valerie

**Comment by vobencha** [comment]
@tiagochst Let me know when I should take another look.
Valerie

**Comment by tiagochst** [comment]
@vobencha I believe now it is ok. I checked the packages and added the missing ones.

**Comment by vobencha** [comment]
OK, I think we're set. Thanks for all the work you've put into this package. I appreciate how receptive you were to suggestions and how quickly you implemented them.

Two things need to happen with this issue, we need to add the workflow svn and Jenkins which @lshep will probably do. @mtmorgan will be adding the data package to svn and grant you permissions on that repo since you already have svn credentials. The workflow may be broken for a few days until the data package propagates through the repositories. 

Valerie

**Comment by lshep** [comment]
I added the workflow to svn and on Jenkins as TCGAWorkflow.  @mtmorgan please grant permissions.

**Comment by tiagochst** [comment]
@vobencha Thank very much for the review. I believe the workflow is much better than before,and I'm sorry it took so long to make all improvements.
